File: app/src/main/java/com/example/ajay/contacts_4/ContactsManager.java
package com.example.ajay.contacts_4;

import android.content.ContentProviderOperation;
import android.provider.ContactsContract.*;
import android.content.ContentProviderResult;
import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.Context;
import android.content.OperationApplicationException;
import android.database.Cursor;
import android.net.Uri;
import android.os.RemoteException;
import android.provider.ContactsContract;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ajay on 28/9/15.
 */
public class ContactsManager {

    private static String MIMETYPE = "vnd.android.cursor.item/com.example.ajay.contacts_4";

    public static void addContact(Context context,MyContact contact){

        ContentResolver resolver = context.getContentResolver();
        boolean mHasAccount = isAlreadyRegistered(resolver, contact.Id);

        if(mHasAccount){
            Log.I(context.getString(R.string.account_exists));
        } else {

            ArrayList<ContentProviderOperation> ops = new ArrayList<ContentProviderOperation>();

            // insert account name and account type
            ops.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.RawContacts.CONTENT_URI, true))
                    .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, Constants.ACCOUNT_NAME)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, Constants.ACCOUNT_TYPE)
                    .withValue(ContactsContract.RawContacts.AGGREGATION_MODE,
                            ContactsContract.RawContacts.AGGREGATION_MODE_DEFAULT)
                    .build());

            // insert contact number
            ops.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                    .withValue(CommonDataKinds.Phone.NUMBER, contact.number)
                    .build());

            // insert contact name
//            ops.add(ContentProviderOperation
//                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
//                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
//                    .withValue(ContactsContract.Data.MIMETYPE,
//                            ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE)
//                    .withValue(CommonDataKinds.StructuredName.DISPLAY_NAME, contact.name)
//                    .build());

            // insert mime-type data
            ops.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, MIMETYPE)
                    .withValue(ContactsContract.Data.DATA1, 12345)
                    .withValue(ContactsContract.Data.DATA2, "user")
                    .withValue(ContactsContract.Data.DATA3, "MyData")
                    .build());

            try {
                resolver.applyBatch(ContactsContract.AUTHORITY, ops);
            } catch (Exception e) {
                e.printStackTrace();
            }

        }
    }

    /**
     * Check if contact is already registered with app
     * @param resolver
     * @param id
     * @return
     */
    private static boolean isAlreadyRegistered(ContentResolver resolver, String id){

        boolean isRegistered = false;
        List<String> str = new ArrayList<>();

        //query raw contact id's from the contact id
        Cursor c = resolver.query(RawContacts.CONTENT_URI, new String[]{RawContacts._ID},
                RawContacts.CONTACT_ID + "=?",
                new String[]{id}, null);

        //fetch all raw contact id's and save them in a list of string
        if(c != null && c.moveToFirst()) {
            do{
                str.add(c.getString(c.getColumnIndexOrThrow(RawContacts._ID)));
            }while (c.moveToNext());
            c.close();
        }

        //query account types and check the account type for each raw contact id
        for(int i=0; i<str.size();i++) {
            Cursor c1 = resolver.query(RawContacts.CONTENT_URI, new String[]{RawContacts.ACCOUNT_TYPE},
                    RawContacts._ID + "=?",
                    new String[]{str.get(i)}, null);

            if(c1 != null){
                c1.moveToFirst();
                String accType = c1.getString(c1.getColumnIndexOrThrow(RawContacts.ACCOUNT_TYPE));
                if(accType != null && accType.equals("com.example.ajay.contacts_4")) {
                    isRegistered = true;
                    break;
                }
                c1.close();
            }
        }

        return isRegistered;
    }

    /**
     * Check for sync call
     * @param uri
     * @param isSyncOperation
     * @return
     */
    private static Uri addCallerIsSyncAdapterParameter(Uri uri, boolean isSyncOperation) {
        if (isSyncOperation) {
            return uri.buildUpon()
                    .appendQueryParameter(ContactsContract.CALLER_IS_SYNCADAPTER, "true")
                    .build();
        }
        return uri;
    }

//    public static void updateMyContact(Context context,String id){
//
//        Cursor ids = context.getContentResolver().query(ContactsContract.Contacts.CONTENT_URI, null,
//                ContactsContract.Contacts._ID + " = ?",
//                new String[]{id}, null);
//
////        int rawContactId = ids.getColumnIndexOrThrow(ContactsContract.RawContacts.);
//
//        ArrayList<ContentProviderOperation> ops = new ArrayList<>();
//        ops.add(ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
//                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID,0)
//                .withValue(ContactsContract.Contacts._ID, id)
//                .withValue(ContactsContract.Data.DATA1, "Data1")
//                .withValue(ContactsContract.Data.DATA2, "Data2")
//                .withValue(ContactsContract.Data.DATA3, "MyData")
//                .build());
//
//        ids.close();
//
//        try{
//            context.getContentResolver().applyBatch(ContactsContract.AUTHORITY,ops);
//        } catch (RemoteException e) {
//            e.printStackTrace();
//        } catch (OperationApplicationException e) {
//            e.printStackTrace();
//        }
//    }
}

File: app/src/main/java/com/ajay/synccontacts/utils/ContactsManager.kt
package com.ajay.synccontacts.utils

import android.content.ContentProviderOperation
import android.content.ContentUris
import android.content.Context
import android.net.Uri
import android.provider.ContactsContract
import android.provider.ContactsContract.AggregationExceptions
import android.provider.ContactsContract.CommonDataKinds
import android.util.Log
import com.ajay.synccontacts.R
import com.ajay.synccontacts.model.Contact

class ContactsManager {

    companion object {

        private const val MESSAGE_MIME_TYPE = "vnd.android.cursor.item/com.ajay.synccontacts.message"
        private const val VOICE_MIME_TYPE = "vnd.android.cursor.item/com.ajay.synccontacts.voice"
        private const val VIDEO_MIME_TYPE = "vnd.android.cursor.item/com.ajay.synccontacts.video"

        /**
         * Method to register a contact with the app
         */
        fun registerContact(context: Context,contact: Contact) {

            for(number in contact.numbers) {
                val operations = ArrayList<ContentProviderOperation>()

                // insert account name and type
                operations.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.RawContacts.CONTENT_URI, true))
                    .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, Constants.ACCOUNT_NAME)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, Constants.ACCOUNT_TYPE)
                    .withValue(ContactsContract.RawContacts.AGGREGATION_MODE,
                        ContactsContract.RawContacts.AGGREGATION_MODE_DEFAULT)
                    .build()
                )

                // insert by phone number (because its unique)
                operations.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                    .withValue(CommonDataKinds.Phone.NUMBER, number) // Supply the number to be synced
                    .build()
                )

                /* (This will be the data retrieved when you click you app from contacts) */
                // insert your app data for message
                operations.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, MESSAGE_MIME_TYPE)
                    .withValue(ContactsContract.Data.DATA1, number)
                    .withValue(ContactsContract.Data.DATA2, contact.name)
                    .withValue(ContactsContract.Data.DATA3, "Message $number")
                    .build()
                )

                // insert your app data for voice call
                operations.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, VOICE_MIME_TYPE)
                    .withValue(ContactsContract.Data.DATA1, number)
                    .withValue(ContactsContract.Data.DATA2, contact.name)
                    .withValue(ContactsContract.Data.DATA3, "Voice call $number")
                    .build()
                )

                // insert your app data for video call
                operations.add(ContentProviderOperation
                    .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, VIDEO_MIME_TYPE)
                    .withValue(ContactsContract.Data.DATA1, number)
                    .withValue(ContactsContract.Data.DATA2, contact.name)
                    .withValue(ContactsContract.Data.DATA3, "Video call $number")
                    .build()
                )

                val contentProviderResult = context.contentResolver.applyBatch(ContactsContract.AUTHORITY, operations)

                /*
                Manually aggregate the new rawcontactid with a rawcontactid of number to be registered
                (If automatic aggregation does not work)
                 */
                val newRawContactId = ContentUris.parseId(contentProviderResult[0].uri).toString()
                Log.d(ContactsManager::class.java.simpleName, "New RawContactId -> $newRawContactId")
                val contactRawContactId = contact.rawContactIdMap[number]
                Log.d(ContactsManager::class.java.simpleName,
                    "RawContactId for number to be registered -> $contactRawContactId")

                if(newRawContactId != null && contactRawContactId != null) {
                    manuallyAggregate(context, newRawContactId, contactRawContactId)
                }
            }
        }

        /**
         * Method to register a number with the app
         */
        fun registerNumber(context: Context,number: String, contactName: String,rawContactIdMap: HashMap<String,String>) {
            val operations = ArrayList<ContentProviderOperation>()

            // insert account name and type
            operations.add(ContentProviderOperation
                .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.RawContacts.CONTENT_URI, true))
                .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, Constants.ACCOUNT_NAME)
                .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, Constants.ACCOUNT_TYPE)
                .withValue(ContactsContract.RawContacts.AGGREGATION_MODE,
                    ContactsContract.RawContacts.AGGREGATION_MODE_DEFAULT)
                .build()
            )

            // insert by phone number (because its unique)
            operations.add(ContentProviderOperation
                .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                .withValue(CommonDataKinds.Phone.NUMBER, number) // Supply the number to be synced
                .build()
            )

            /* (This will be the data retrieved when you click you app from contacts) */
            // insert your app data for message
            operations.add(ContentProviderOperation
                .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, MESSAGE_MIME_TYPE)
                .withValue(ContactsContract.Data.DATA1, number)
                .withValue(ContactsContract.Data.DATA2, contactName)
                .withValue(ContactsContract.Data.DATA3, "Message $number")
                .build()
            )

            // insert your app data for voice call
            operations.add(ContentProviderOperation
                .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, VOICE_MIME_TYPE)
                .withValue(ContactsContract.Data.DATA1, number)
                .withValue(ContactsContract.Data.DATA2, contactName)
                .withValue(ContactsContract.Data.DATA3, "Voice call $number")
                .build()
            )

            // insert your app data for video call
            operations.add(ContentProviderOperation
                .newInsert(addCallerIsSyncAdapterParameter(ContactsContract.Data.CONTENT_URI, true))
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, VIDEO_MIME_TYPE)
                .withValue(ContactsContract.Data.DATA1, number)
                .withValue(ContactsContract.Data.DATA2, contactName)
                .withValue(ContactsContract.Data.DATA3, "Video call $number")
                .build()
            )

            val contentProviderResult = context.contentResolver.applyBatch(ContactsContract.AUTHORITY, operations)

            /*
            Manually aggregate the new rawcontactid with a rawcontactid of number to be registered
            (If automatic aggregation does not work)
             */
            val newRawContactId = ContentUris.parseId(contentProviderResult[0].uri).toString()
            Log.d(ContactsManager::class.java.simpleName, "New RawContactId -> $newRawContactId")
            val contactRawContactId = rawContactIdMap[number]
            Log.d(ContactsManager::class.java.simpleName,
                "RawContactId for number to be registered -> $contactRawContactId")

            if(newRawContactId != null && contactRawContactId != null) {
                manuallyAggregate(context, newRawContactId, contactRawContactId)
            }
        }

        /**
         * Method to delete RawContact of number specified
         */
        fun deleteNumber(context: Context,number: String) {
            //region Get RawContactId's for the number
            val rawContactIdCursor = context.contentResolver.query(CommonDataKinds.Phone.CONTENT_URI,
                arrayOf(CommonDataKinds.Phone.RAW_CONTACT_ID),
                "${CommonDataKinds.Phone.NUMBER} = ?",
                arrayOf(number), null)

            val rawContactIdList = ArrayList<String>()
            if(rawContactIdCursor != null && rawContactIdCursor.moveToFirst()) {
                do{
                    rawContactIdList.add(rawContactIdCursor.getString(rawContactIdCursor.getColumnIndexOrThrow(
                        CommonDataKinds.Phone.RAW_CONTACT_ID
                    )))
                }while (rawContactIdCursor.moveToNext())
                rawContactIdCursor.close()
            }
            //endregion

            //region Get the RawContactId with the app's account type
            var appRawContactId = ""
            for(rawContactId in rawContactIdList) {
                val accTypeCursor = context.contentResolver.query(ContactsContract.RawContacts.CONTENT_URI,
                    arrayOf(ContactsContract.RawContacts.ACCOUNT_TYPE),
                    "${ContactsContract.RawContacts._ID} = ?",
                    arrayOf(rawContactId), null
                )

                if(accTypeCursor != null && accTypeCursor.moveToFirst()) {
                    do{
                        val accountType = accTypeCursor.getString(accTypeCursor.getColumnIndexOrThrow(
                            ContactsContract.RawContacts.ACCOUNT_TYPE
                        ))
                        if(accountType == Constants.ACCOUNT_TYPE) {
                            appRawContactId = rawContactId
                        }
                    }while (accTypeCursor.moveToNext())
                    accTypeCursor.close()
                }

                if(appRawContactId.isNotEmpty())
                    break
            }
            //endregion

            //region Perform delete
            val operations = ArrayList<ContentProviderOperation>()

            operations.add(ContentProviderOperation
                .newDelete(ContactsContract.RawContacts.CONTENT_URI)
                .withSelection("${ContactsContract.RawContacts._ID} = ?", arrayOf(appRawContactId))
                .build()
            )

            context.contentResolver.applyBatch(ContactsContract.AUTHORITY, operations)
            //endregion
        }

        /**
         * Method to manually aggregate two raw contacts
         */
        private fun manuallyAggregate(context: Context,id1: String,id2: String) {
            val ops = ArrayList<ContentProviderOperation>()

            ops.add(ContentProviderOperation
                .newUpdate(AggregationExceptions.CONTENT_URI)
                .withValue(AggregationExceptions.TYPE, AggregationExceptions.TYPE_KEEP_TOGETHER)
                .withValue(AggregationExceptions.RAW_CONTACT_ID1, id1)
                .withValue(AggregationExceptions.RAW_CONTACT_ID2, id2)
                .build()
            )

            context.contentResolver.applyBatch(ContactsContract.AUTHORITY, ops)
        }

        /**
         * Check for sync enabled or disabled
         */
        private fun addCallerIsSyncAdapterParameter(uri: Uri, isSyncOperation: Boolean): Uri {
            return if(isSyncOperation){
                uri.buildUpon()
                    .appendQueryParameter(ContactsContract.CALLER_IS_SYNCADAPTER, "true")
                    .build()
            } else {
                uri
            }
        }
    }
}
